src/lib/integration_test_lib/event_type.ml

```diff
@@ -417,35 +427,89 @@ let event_to_yojson event =
+  match Ty.parse with From_daemon_log (id, _) -> Some id | _ -> None
 
-let of_structured_event_id =
+let structured_events_table =
   let open Option.Let_syntax in
-  let table =
-    all_event_types
-    |> List.filter_map ~f:(fun t ->
-           let%map event_id = to_structured_event_id t in
-           (Structured_log_events.string_of_id event_id, t))
-    |> String.Table.of_alist_exn
-  in
-  Fn.compose (String.Table.find table) Structured_log_events.string_of_id
-
-let parse_event (message : Logger.Message.t) =
+  all_event_types
+  |> List.filter_map ~f:(fun t ->
+         let%map event_id = to_structured_event_id t in
+         (Structured_log_events.string_of_id event_id, t))
+  |> String.Table.of_alist_exn
+
+let of_structured_event_id id =
+  Structured_log_events.string_of_id id
+  |> String.Table.find structured_events_table
+
+let to_puppeteer_event_string event_type =
+  let (Event_type t) = event_type in
+  let (module Ty) = event_type_module t in
+  match Ty.parse with From_puppeteer_log (id, _) -> Some id | _ -> None
+
+let puppeteer_events_table =
+  let open Option.Let_syntax in
+  all_event_types
+  |> List.filter_map ~f:(fun t ->
+         let%map event_id = to_puppeteer_event_string t in
+         (event_id, t))
+  |> String.Table.of_alist_exn
+
+let of_puppeteer_event_string id = String.Table.find puppeteer_events_table id
+
+let parse_daemon_log (message : Logger.Message.t) =
   let open Or_error.Let_syntax in
   match message.event_id with
-  | Some event_id ->
-      let (Event_type event_type) =
+  | Some event_id -> (
+      let (Event_type ev_type) =
         of_structured_event_id event_id
         |> Option.value ~default:(Event_type Log_error)
       in
-      let (module Type) = event_type_module event_type in
-      let%map data = Type.parse message in
-      Event (event_type, data)
-  | None ->
+      let (module Ty) = event_type_module ev_type in
+      match Ty.parse with
+      | From_error_log fnc | From_daemon_log (_, fnc) ->
+          let%map data = fnc message in
+          Event (ev_type, data)
+      | _ ->
+          failwith
+            "the 'parse' field of a daemon event should always be a \
+             `From_daemon_log variant, but was mis-programmed as something \
+             else. this is a progammer error" )
+  | None -> (
       (* TODO: check log level to ensure it matches error log level *)
-      let%map data = Log_error.parse message in
-      Event (Log_error, data)
+      match Log_error.parse with
+      | From_error_log fnc ->
+          let%map data = fnc message in
+          Event (Log_error, data)
+      | _ ->
+          failwith
+            "Log_error.parse should always be a `From_error_log variant, but \
+             was mis-programmed as something else. this is a progammer error" )
+
+let parse_puppeteer_log (message : Puppeteer_message.t) =
+  let open Or_error.Let_syntax in
+  match
+    Option.( >>= ) message.puppeteer_event_type of_puppeteer_event_string
```

Comment:
You can also write this as `Option.bind message.puppeteer_event_type of_puppeteer_event_string` or `Option.(message.puppeteer_event_type >>= of_puppeteer_event_string)`